src/lib.rs

```diff
@@ -0,0 +1,497 @@
+use std::{collections::HashMap, convert::TryFrom, io::Write, str::FromStr};
+
+use anyhow::{Context, Result};
+use log::info;
+use mbedtls::pk::Pk;
+pub use sdkms::api_model::ObjectType;
+use sdkms::{
+    api_model::{
+        KeyOperations, RsaEncryptionPaddingPolicy, RsaEncryptionPolicy,
+        RsaOptions, RsaSignaturePaddingPolicy, RsaSignaturePolicy, Sobject,
+        SobjectDescriptor, SobjectRequest,
+    },
+    SdkmsClient,
+};
+extern crate sequoia_openpgp as openpgp;
+use openpgp::{
+    crypto::SessionKey,
+    packet::{
+        key::{
+            Key4, PrimaryRole, PublicParts, SubordinateRole, UnspecifiedRole,
+        },
+        signature::SignatureBuilder,
+        Key, UserID, PKESK, SKESK,
+    },
+    parse::{
+        stream::{
+            DecryptionHelper, DecryptorBuilder, MessageStructure,
+            VerificationHelper,
+        },
+        Parse,
+    },
+    policy::Policy as PgpPolicy,
+    serialize::{
+        stream::{Armorer, Message, Signer},
+        SerializeInto,
+    },
+    types::{HashAlgorithm, KeyFlags, SignatureType, SymmetricAlgorithm},
+    Cert, KeyHandle, Packet,
+};
+use uuid::Uuid;
+
+mod signer;
+
+mod decryptor;
+
+use decryptor::RawDecryptor;
+use signer::RawSigner;
+
+/// The PgpAgent is responsible for one PGP key. It talks to SDKMS and produces
+/// valid OpenPGP material
+pub struct PgpAgent {
+    api_endpoint:    String,
+    api_key:         String,
+    primary:         PublicKey,
+    subkey:          PublicKey,
+    pub certificate: Option<Cert>,
+}
+
+#[derive(Clone)]
+struct PublicKey {
+    role:        KeyRole,
+    descriptor:  SobjectDescriptor,
+    sequoia_key: Option<Key<PublicParts, UnspecifiedRole>>,
+}
+
+#[derive(Clone)]
+enum KeyRole {
+    Primary,
+    Subkey,
+}
+
+pub enum SupportedPkAlgo {
+    Rsa(u32),
+}
+
+impl PgpAgent {
+    /// Generates an OpenPGP key with secrets stored in SDKMS. At the OpenPGP
+    /// level, this method produces a key consisting of
+    ///
+    /// - A primary signing key (flags C + S),
+    /// - a transport and storage encryption subkey (flags Et + Er).
+    ///
+    /// At the SDKMS level, this method creates the two corresponding Sobjects.
+    /// The encryption key's KID is stored as a custom metadata field of the
+    /// signing key.
+    ///
+    /// The public certificate (Transferable Public Key) is computed, stored as
+    /// an additional custom medatada field on the primary key, and returned.
+    pub fn generate_key(
+        api_endpoint: &str,
+        api_key: &str,
```

Comment:
Have you considered taken an `SdkmsClient` instance here and in other methods that currently accept an `api_endpoint` and `api_key` (`PgpAgent` would store this instance as a member instead of the `api_endpoint` and `api_key` fields)? This would allow clients of this struct to initialize an `SdkmsClient` once and have the `PgpAgent` reuse that instance whenever needed. It would also decouple the `PgpAgent` from the specific way that an `SdkmsClient` is created.

Reply:
This would be cleaner, I agree - it will also allow to control the lifetime of the `SdkmsClient`. I actually tried this on my first implementation, but I ran into some problems when implementing the `Signer` trait (IIRC, problems with the `Sync` trait within `SdkmsClient`). I'll give it another shot.

Reply:
Right so as soon as `RawSigner` has a `http_client: &'a SdkmsClient` field, the `Signer` implementation fails with
```
error[E0277]: `RefCell<u64>` cannot be shared between threads safely
   --> src/lib.rs:138:27
    |
138 |         let mut message = Signer::new(message, signer).detached().build()?;
    |                           ^^^^^^^^^^^ `RefCell<u64>` cannot be shared between threads safely
    | 
   ::: /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/sequoia-openpgp-1.1.0/src/serialize/stream.rs:743:35
    |
743 |         where S: crypto::Signer + Send + Sync + 'a
    |                                   ---- required by this bound in `sequoia_openpgp::serialize::stream::Signer::<'a>::new`
    |
    = help: within `SdkmsClient`, the trait `Sync` is not implemented for `RefCell<u64>`
    = note: required because it appears within the type `SdkmsClient`
    = note: required because of the requirements on the impl of `Send` for `&SdkmsClient`
    = note: required because it appears within the type `RawSigner<'_>`

```
Any thoughts?

Reply:
From the error message, `RefCell<T>` is not sync in any case, I believe you could to use a `Arc<RefCell<T>>`.

Reply:
@janosimas thanks. This `RefCell<u64>` is a field from the `SdkmsClient` [(repo)](https://github.com/fortanix/sdkms-client-rust) so there's not much we can do here? (Any workaround will sooner or later `Sync` that `RefCell` field...)

If I add `unsafe impl Sync for SdkmsClient {}` (in the `sdkms` crate) everything works though <- But I'm not sure of the implications of this

Reply:
I'm no specialist here but the `Sync` trait is just a marker, it states that your type is safe to share references between threads.
Since it has a field that is not `Sync` (the `RefCell`) than it should also not be `Sync`.
If the `Sync` trait is a requirement, we should think about changing the `SdkmsClient`.

Reply:
:+1: yes, it is a requirement for Sequoia's `Signer` trait.

Here is the working code: https://github.com/fortanix/sq-sdkms/pull/20 (if `SdkmsClient` was `Sync`).

Reply:
`SdkmsClient` indeed isn't `Sync` (so we certainly can't add `unsafe impl Sync for SdkmsClient`) but Sequoia's `Signer` struct requires its inner signers (which contain an `&SdkmsClient` in PR #20) to be `Send` and `Sync`. I think for our purpose it would be okay if the `Signer` struct wasn't `Sync`, but Sequoia doesn't provide an `!Sync` variant.

For the sake of argument, what would probably work is to have `PgpAgent` contain an `SdkmsClient` and have `RawSigner` contain a `Mutex<SdkmsClient>`, but I don't think we want to go that route though.

Another option would be to have `PgpAgent` contain an `SdkmsBuilder`. That actually closely resembles the semantics of passing around the API endpoint and key, because it means that, although `PgpAgent` doesn't contain an actual `SdkmsClient`, it has a way of constructing one when needed. Unfortunately, `SdkmsBuilder` doens't appear to implement `Clone`, so we'll run into issues here as well.

Probably it makes most sense to stick to the original approach of passing around the API endpoint and key.

Reply:
Interesting, thanks for pointing out these alternatives. I'll see if something can be made about this in the future, in #20